cpp/serve/model.h

```diff
@@ -91,6 +96,58 @@ class ModelObj : public Object {
    */
   virtual ObjectRef ImageEmbed(const NDArray& image, ObjectRef* dst = nullptr, int offset = 0) = 0;
 
+  /*!
+   * \brief Fuse the embeddings and hidden_states.
+   * \param embeddings The embedding of the input to be prefilled.
+   * \param hidden_states_in The hidden_states from previous base model.
+   * \param batch_size Batch size.
+   * \param seq_len Sequence length.
+   * \return The fused hidden_staets.
+   */
+  virtual NDArray FuseEmbedHidden(const ObjectRef& embeddings, const ObjectRef& hidden_states_in,
+                                  int batch_size, int seq_len) = 0;
+
+  /*!
+   * \brief Return if the model has head.
+   */
+  virtual bool CanGetLogits() = 0;
+
+  /*!
+   * \brief Compute logits for hidden_states.
+   * \param hidden_states_in The hidden_states to compute logits for.
+   * \return The computed logits.
+   */
+  virtual NDArray GetLogits(const ObjectRef& hidden_states_in) = 0;
+
+  /*!
+   * \brief Compute logits for hidden_states in a batch.
+   * \param hidden_states_in The hidden_states to compute logits for.
+   * \param seq_ids The id of the sequence in the KV cache.
+   * \param lengths The length of each sequence to prefill.
+   * \return The computed logits.
+   */
+  virtual NDArray BatchGetLogits(const ObjectRef& hidden_states_in,
+                                 const std::vector<int64_t>& seq_ids,
+                                 const std::vector<int>& lengths) = 0;
+
+  /*!
+   * \brief Select last hidden_states for hidden_states in a batch.
+   * \param hidden_states_in The hidden_states to compute logits for.
+   * \param seq_ids The id of the sequence in the KV cache.
+   * \param lengths The length of each sequence to prefill.
+   * \return The last hidden_states for the batch.
+   */
+  virtual NDArray BatchGetHidden(const ObjectRef& hidden_states_in,
+                                 const std::vector<int64_t>& seq_ids,
+                                 const std::vector<int>& lengths) = 0;
+
+  /*!
+   * \brief Concat a list of 1D hidden_states to 2D tensor.
+   * \param hidden_states The hidden_states to concat.
+   * \param dst The copy destination.
+   */
+  virtual NDArray ConcatHidden(std::vector<NDArray>& hidden_states, ObjectRef* dst) = 0;
```

Comment:
Just to confirm: is the input `std::vector<NDArray>` rather than `std::vector<ObjectRef>`, given they might be DRef when tensor parallelism is enabled?

Reply:
I am not familiar with tensor parallelism settings. Not quite sure if we need to use `ObjectRef`.